backends/ent/annotations.go

```diff
@@ -14,11 +14,12 @@ import (
 	"github.com/protobom/storage/internal/backends/ent"
 	"github.com/protobom/storage/internal/backends/ent/annotation"
 	"github.com/protobom/storage/internal/backends/ent/document"
+	"github.com/protobom/storage/internal/backends/ent/node"
 	"github.com/protobom/storage/internal/backends/ent/predicate"
 )
 
-// AddAnnotations applies multiple named annotation values to a single document.
-func (backend *Backend) AddAnnotations(documentID, name string, values ...string) error {
+// AddDocumentAnnotations applies multiple named annotation values to a single document.
+func (backend *Backend) AddDocumentAnnotations(documentID, name string, values ...string) error {
```

Comment:
Not necessarily a good idea, but just an example of what generalizing the logic using a generic function might look like. It uses:

- a constrained type parameter (`[T *sbom.Document | *sbom.Node]`)
- a type switch that can handle either document or node ID

The backend must be passed in since we can't implement this as a method with a receiver.

```go
func AddAnnotations[T *sbom.Document | *sbom.Node](backend *Backend, id, name string, values ...string) error {
	var typ T

	switch any(typ).(type) {
	case *sbom.Document:
		// add annotations to Document
	case *sbom.Node:
		// add annotations to Node
	}

	return nil
}
```

Since the type can't be inferred from its usage in the function params or return type, it would need to be called like:

```go
func main() {
	if err := ent.AddAnnotations[*sbom.Document]("<id>", "<name>"); err != nil {
		panic(err)
	}

	if err := ent.AddAnnotations[*sbom.Node]("<id>", "<name>"); err != nil {
		panic(err)
	}

	// This doesn't even compile since it doesn't satisfy the constrained type parameter 
	// so no need for a `default:` case in the type switch
	if err := ent.AddAnnotations[*sbom.Metadata]("<id>", "<name>"); err != nil {
		panic(err)
	}
}
```

Another option might be to revert back to the original name `backend.AddAnnotations`, and add a `typ any` parameter, then do type assertions to check if it's a `Node` or a `Document`

```go
func (backend *Backend) AddAnnotations(typ any, id, name string, values ...string) error {
	if document, ok := typ.(*sbom.Document); ok {
		// add annotations to Document
	}

	if node, ok := typ.(*sbom.Node); ok {
		// add annotations to Node
	}

	return fmt.Errorf("unsupported type: %T", typ)
}
```

Reply:
These are just comments/examples based on our earlier discussion, not a blocking thread

Reply:
Not crazy about either of these solutions right off the bat. It seems to me like using separate methods for nodes and documents is a bit more straightforward on the user side.